src/coreclr/vm/interoplibinterface_shared.cpp

```diff
@@ -22,9 +25,32 @@ bool Interop::ShouldCheckForPendingException(_In_ NDirectMethodDesc* md)
     PTR_CUTF8 libraryName = md->GetLibNameRaw();
     PTR_CUTF8 entrypointName = md->GetEntrypointName();
     if (libraryName == NULL || entrypointName == NULL)
-        return false;
+    {
+        // If no sig info was supplied we have no recourse.
+        if (sigInfo == NULL)
+            return false;
+
+        if (libraryName == NULL)
+        {
+            // If the library name is unknown, see if we can get the token
+            // from the sig info.
```

Comment:
I think it would be better to require `NDirectMethodDesc` to be populated when this is called instead of trying to patch it up here.

When this gets called from the exception handler, I think it should be safe to assume that `NDirectMethodDesc` has been populated (in all cases that matter).

Reply:
>When this gets called from the exception handler

This isn't the issue actually. The issue is during jit compilation when we are trying to determine if the P/Invoke call can be inlined. When the P/Invoke is eventually invoked, the PreStub path fills in the `NDirectMethodDesc` and since that has a conceptual lock around it the update is safe there but not at this point. I can try to audit the path but I do fear the potential for bug tail.

Reply:
`PopulateNDirectMethodDesc` seems to be thread-safe already, it is just not very efficient when called multiple times.

Reply:
My concern is the following comment:

https://github.com/dotnet/runtime/blob/64f5f0b31d9af2df199229e6e4ef3f4a876a05a5/src/coreclr/vm/dllimport.cpp#L4266-L4268

Reply:
I think that this comment was meant to say that the flags have to be set atomically.

It is pretty easy to see that `PopulateNDirectMethodDesc` can be called multiple threads at the same time today.